### PR TITLE
[SPARK-40975][SQL] Rename the error class `_LEGACY_ERROR_TEMP_0021` to `UNSUPPORTED_TYPED_LITERAL`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to assign the proper name `UNSUPPORTED_TYPED_LITERAL ` to the legacy error class `_LEGACY_ERROR_TEMP_0021 `, and modify test suite to use `checkError()` which checks the error class name, context and etc.

### Why are the changes needed?
Proper name improves user experience w/ Spark SQL.

### Does this PR introduce _any_ user-facing change?
Yes, the PR changes an user-facing error message. 

### How was this patch tested?
By running the modified test suites:
```
$ PYSPARK_PYTHON=python3 build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite"
$ build/sbt "test:testOnly *ExpressionParserSuite"
$ build/sbt "test:testOnly *AnalysisSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
The failure is not related to my changes:
```
ERROR: test_await_termination_or_timeout (pyspark.streaming.tests.test_context.StreamingContextTests)
```
The same finished successfully on my laptop.

**Comment by MaxGekk** [comment]
@cloud-fan @itholic @srielau @LuciferYang @panbingkun Please, review this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -2397,7 +2397,11 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
                 s"contains illegal character for hexBinary: $padding$value");
           }
         case other =>
-          throw QueryParsingErrors.literalValueTypeUnsupportedError(other, ctx)
+          throw QueryParsingErrors.literalValueTypeUnsupportedError(
+            unsupportedType = other,
+            supportedTypes =
+              Seq("DATE", "TIMESTAMP_NTZ", "TIMESTAMP_LTZ", "TIMESTAMP", "INTERVAL", "X"),
```
Comment (srielau) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2403]: What's X? Hexadecimal?

**Review by srielau — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2403]: Yep

Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2403]: Maybe a lowercase `hexadecimal` would be less confusing?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2403]: Here, I pointed out the exact keyword that can be used in the typed literals, see
https://github.com/apache/spark/blob/933dc0c42f0caf74aaa077fd4f2c2e7208452b9b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L2390

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2403]: We could give a hint to user like `X hexadecimal`

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2403]: Do you mean `"X"` -> `"X hexadecimal"` ?

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2403]: Agree with you, `"X hexadecimal"` looks a little strange

**Review by LuciferYang — APPROVED** [review]
+1, fine to me

**Review by itholic — APPROVED** [review]
LGTM when the test passing

File: sql/core/src/test/resources/sql-tests/results/literals.sql.out [hunk 2]
```diff
@@ -442,9 +442,10 @@ struct<>
 -- !query output
 org.apache.spark.sql.catalyst.parser.ParseException
 {
-  "errorClass" : "_LEGACY_ERROR_TEMP_0021",
+  "errorClass" : "UNSUPPORTED_TYPED_LITERAL",
   "messageParameters" : {
-    "valueType" : "GEO"
+    "supportedTypes" : "\"DATE\", \"TIMESTAMP_NTZ\", \"TIMESTAMP_LTZ\", \"TIMESTAMP\", \"INTERVAL\", \"X\"",
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/literals.sql.out:447]: isn't it a constant list? does it need to be a parameter?

Comment (MaxGekk) [reply sql/core/src/test/resources/sql-tests/results/literals.sql.out:447]: It depends on implementation, and we might support more in the future. Having the list in the source code can give the following benefits:
1. More likely, devs will not forget to update the list in source code.
2. If the list will be in the JSON file, tech editors might forget to edit it or accidentally modify it.
3. The list of parameters might be formatted in different ways ( as a foldable or drop-down list) by frontend tools.

Comment (LuciferYang) [reply sql/core/src/test/resources/sql-tests/results/literals.sql.out:447]: I think it should be a parameter, unless `UNSUPPORTED_TYPED_LITERAL` will only be used in this scenario

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @cloud-fan @LuciferYang @itholic @srielau for review.

**Pull request closed without merging**